### PR TITLE
Chapter1

### PR DESCRIPTION
**Pull request by DamjanOstrelic**
### Solutions for Chapter 1

cc @vrom911 @chshersh

Hello, creators of this very cool course! Am I submitting my solutions the right way? PR into the `main` branch?
Would love some feedback from you :) Thanks for creating this content.

I had troubles running tests locally when running `make test-chapter1` (error log below), but tests passed on the GitHub CI.

```
Configuring library for syb-0.7.1..
Preprocessing library for syb-0.7.1..
Building library for syb-0.7.1..
[ 1 of 16] Compiling Data.Generics.Aliases ( src/Data/Generics/Aliases.hs, dist/build/Data/Generics/Aliases.o, dist/build/Data/Generics/Aliases.dyn_o )
[ 2 of 16] Compiling Data.Generics.Basics ( src/Data/Generics/Basics.hs, dist/build/Data/Generics/Basics.o, dist/build/Data/Generics/Basics.dyn_o )
[ 3 of 16] Compiling Data.Generics.Builders ( src/Data/Generics/Builders.hs, dist/build/Data/Generics/Builders.o, dist/build/Data/Generics/Builders.dyn_o )
[ 4 of 16] Compiling Data.Generics.Instances ( src/Data/Generics/Instances.hs, dist/build/Data/Generics/Instances.o, dist/build/Data/Generics/Instances.dyn_o )
[ 5 of 16] Compiling Data.Generics.Schemes ( src/Data/Generics/Schemes.hs, dist/build/Data/Generics/Schemes.o, dist/build/Data/Generics/Schemes.dyn_o )
[ 6 of 16] Compiling Data.Generics.Text ( src/Data/Generics/Text.hs, dist/build/Data/Generics/Text.o, dist/build/Data/Generics/Text.dyn_o )
[ 7 of 16] Compiling Data.Generics.Twins ( src/Data/Generics/Twins.hs, dist/build/Data/Generics/Twins.o, dist/build/Data/Generics/Twins.dyn_o )
[ 8 of 16] Compiling Data.Generics    ( src/Data/Generics.hs, dist/build/Data/Generics.o, dist/build/Data/Generics.dyn_o )
[ 9 of 16] Compiling Generics.SYB     ( src/Generics/SYB.hs, dist/build/Generics/SYB.o, dist/build/Generics/SYB.dyn_o )
[10 of 16] Compiling Generics.SYB.Aliases ( src/Generics/SYB/Aliases.hs, dist/build/Generics/SYB/Aliases.o, dist/build/Generics/SYB/Aliases.dyn_o )
[11 of 16] Compiling Generics.SYB.Basics ( src/Generics/SYB/Basics.hs, dist/build/Generics/SYB/Basics.o, dist/build/Generics/SYB/Basics.dyn_o )
[12 of 16] Compiling Generics.SYB.Builders ( src/Generics/SYB/Builders.hs, dist/build/Generics/SYB/Builders.o, dist/build/Generics/SYB/Builders.dyn_o )
[13 of 16] Compiling Generics.SYB.Instances ( src/Generics/SYB/Instances.hs, dist/build/Generics/SYB/Instances.o, dist/build/Generics/SYB/Instances.dyn_o )
[14 of 16] Compiling Generics.SYB.Schemes ( src/Generics/SYB/Schemes.hs, dist/build/Generics/SYB/Schemes.o, dist/build/Generics/SYB/Schemes.dyn_o )
[15 of 16] Compiling Generics.SYB.Text ( src/Generics/SYB/Text.hs, dist/build/Generics/SYB/Text.o, dist/build/Generics/SYB/Text.dyn_o )
[16 of 16] Compiling Generics.SYB.Twins ( src/Generics/SYB/Twins.hs, dist/build/Generics/SYB/Twins.o, dist/build/Generics/SYB/Twins.dyn_o )
/usr/bin/ld.gold: error: cannot find -lgmp
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)

```

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -489,8 +489,9 @@ Implement a function that returns the last digit of a given number.
   results. Or you can try to guess the function name, search for it and check
   whether it works for you!
 -}
--- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Int -> Int -- What about last digit of a float?
+lastDigit n = 
+    if n < 0 then mod (n * (-1)) 10 else mod n 10
```
Comment (vrom911) [inline src/Chapter1.hs:494]: You can use `abs` function to get the absolute value of the number :) So you can skip the check on negative numbers 🙂

File: src/Chapter1.hs [hunk 2]
```diff
@@ -553,9 +555,12 @@ value after "=" where the condition is true.
...
 -}
-
-mid x y z = error "mid: not implemented!"
-
+mid :: Int -> Int -> Int -> Int
+mid x y z
+    | (x > y && x < z) || (x > z && x < y) = x
+    | (y > x && y < z) || (y > z && y < x) = y
+    | (z > x && z < y) || (z > y && z < x) = z
+    | otherwise = if x == y then x else z
```
Comment (vrom911) [inline src/Chapter1.hs:563]: I guess if you use `<=` `>=` in the above checks, you won't need the last check here, and could get rid of the third clause as well (and return `z` straightaway) 👌🏼

File: src/Chapter1.hs [hunk 3]
```diff
@@ -631,8 +643,12 @@ Implement a function that returns the sum of the last two digits of a number.
 Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
-
-sumLast2 n = error "sumLast2: Not implemented!"
+sumLast2 :: Int -> Int
+sumLast2 n =
+    let npositive = if n < 0 then n * (-1) else n
```
Comment (vrom911) [inline src/Chapter1.hs:648]: You could use `abs` on `n` here as well 🙂

File: src/Chapter1.hs [hunk 4]
```diff
@@ -631,8 +643,12 @@ Implement a function that returns the sum of the last two digits of a number.
...
+        first = mod npositive 10
+        second = mod (div npositive 10) 10
```
Comment (vrom911) [inline src/Chapter1.hs:650]: That is a wonderful solution! 👏🏼  You correctly noticed that it is the `div` and `mod`, cool 😎 

One hint to make your solution even shorter: you can see that you use both:

```haskell
mod (abs n) 10
div (abs n) 10
```
The standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(x, y) = divMod (abs n) 10
```

You can see how we could pattern match on the pair 🙂

File: src/Chapter1.hs [hunk 5]
```diff
@@ -631,8 +643,12 @@ Implement a function that returns the sum of the last two digits of a number.
...
+    in (first + second)
```
Comment (vrom911) [inline src/Chapter1.hs:651]: You don't need parenthesis here 🙂 
Also, first and second are the names of standard functions, so maybe you could name these vars more precisely to omit conflicts 👌🏼

File: src/Chapter1.hs [hunk 6]
```diff
@@ -652,9 +668,13 @@ Implement a function that returns the first digit of a given number.
...
-
-firstDigit n = error "firstDigit: Not implemented!"
-
+firstDigit :: Int -> Int
+firstDigit n
+    | npositive < 10 = npositive
+    | otherwise = 
+        let x = div npositive 10
+        in (firstDigit x)
+  where npositive = if n < 0 then n * (-1) else n
```
Comment (vrom911) [inline src/Chapter1.hs:677]: It is better to use `abs` function in here as well 🙂 
Also, you don't need parenthesis after `in` keyword.

**Review by vrom911 — APPROVED** [review]
Great work on the Chapter!
Congrats with finishing all task including the advance one 👏🏼 🎉 

Regarding the PR, we advice to open them against your own fork, so it would be easier for you to work. We won't be able to merge your solution to our repo, but you can merge it to your fork, and then the work on opening PRs to other Chapters will be easier, at least because you won't need to handle how to isolate each chapter changes separately 🙂 

Your issue seems to be a linker issue 🤔 Could you please provide more details on what platform you are, what GHC and Cabal version you are using?

```shell
ghc --version
cabal --version
```

File: src/Chapter1.hs (hunk 4 above)
Comment (DamjanOstrelic) [reply src/Chapter1.hs:650]: I see, so is there a simpler way than:

```
sumLast2 :: Int -> Int
sumLast2 n =
    let npositive = abs (n)
        (a, x) = divMod npositive 10
        y = mod a 10
    in x + y
```

**Comment by DamjanOstrelic** [comment]
Thanks for the comments @vrom911! Really helpful tips 👍 

Info on the GHC/Cabal versions:

```
>> ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.10.2
>> cabal --version
cabal-install version 3.2.0.0
compiled using version 3.2.0.0 of the Cabal library 
```

**Comment by vrom911** [comment]
If you are on MacOS this could be the latest Big Sur this could be known issue in there. I am not sure what OS you are using though, so can't say more...

**Comment by vrom911** [comment]
I am going to close this PR as we can't merge it into the course 🙂 
Btw, if you want to discuss something, you can join us in the [Discussions](https://github.com/kowainik/learn4haskell/discussions) section now 👍🏼

**Pull request closed without merging**